Allow NLA for AD desktops

probakowski opened the pull request:
Closes #8546

changelog: Add support for NLA for Windows desktops connected to Active Directory domain.

probakowski committed:
initial

probakowski committed:
full flow

probakowski committed:
e

probakowski committed:
Merge branch 'master' into probakowski/nla

probakowski committed:
cleanup

probakowski committed:
cleanup, deps

probakowski committed:
fix non-ad

probakowski committed:
update IronRDP/sspi, remove domain

probakowski committed:
rework url parsing

probakowski committed:
lint

probakowski committed:
cleanup

probakowski committed:
Merge branch 'master' into probakowski/nla

zmb3 commented on lib/srv/desktop/rdp/rdpclient/src/network_client.rs:
```diff
@@ -0,0 +1,149 @@
+        let addr = format!(
+            "{}:{}",
+            url.host_str().unwrap_or_default(),
+            url.port().unwrap_or(88)
```
```suggestion
            url.port().unwrap_or(DEFAULT_KERBEROS_PORT)
```

zmb3 commented on lib/srv/desktop/rdp/rdpclient/src/network_client.rs:
```diff
@@ -0,0 +1,149 @@
+        let mut stream = TcpStream::connect(addr)
+            .await
+            .map_err(|e| Error::new(ErrorKind::NoAuthenticatingAuthority, format!("{:?}", e)))
+            .map_err(|e| custom_err!("failed to send KDC request over TCP", e))?;
```
```suggestion
            .map_err(|e| custom_err!("failed to connect to KDC", e))?;
```

probakowski replied:
I've added separate error messages

zmb3 commented on lib/srv/desktop/rdp/rdpclient/src/network_client.rs:
```diff
@@ -0,0 +1,149 @@
+            .read_u32()
+            .await
+            .map_err(|e| Error::new(ErrorKind::NoAuthenticatingAuthority, format!("{:?}", e)))
+            .map_err(|e| custom_err!("failed to send KDC request over TCP", e))?;
```
```suggestion
            .map_err(|e| custom_err!("failed to read KDC response over TCP", e))?;
```

probakowski replied:
I've added separate error messages

zmb3 commented on lib/srv/desktop/rdp/rdpclient/src/network_client.rs:
```diff
@@ -0,0 +1,149 @@
+            .await
+            .map_err(|e| custom_err!("failed to send UDP request", e))?;
+
+        // 48 000 bytes: default maximum token len in Windows
+        let mut buf = vec![0; 0xbb80];
```
Make a constant for this value, and ideally link to some MSFT docs that describe it.

zmb3 commented on lib/srv/desktop/rdp/rdpclient/src/network_client.rs:
```diff
@@ -0,0 +1,149 @@
+            .read_exact(&mut buf[4..])
+            .await
+            .map_err(|e| Error::new(ErrorKind::NoAuthenticatingAuthority, format!("{:?}", e)))
+            .map_err(|e| custom_err!("failed to send KDC request over TCP", e))?;
```
Try not to use the same error message in multiple places, it makes troubleshooting difficult.

probakowski replied:
I've added separate error messages

zmb3 commented on lib/srv/desktop/windows_server.go:
```diff
@@ -858,6 +861,17 @@ func (s *WindowsService) connectRDP(ctx context.Context, log *slog.Logger, tdpCo
+
+	computerName, ok := desktop.GetLabel(types.DiscoveryLabelWindowsDNSHostName)
+	if !ok {
+		computerName = strings.Split(desktop.GetAddr(), ":")[0]
```
Use `net.SplitHostPort` or one of our internal utils instead?

zmb3 commented on lib/srv/desktop/windows_server.go:
```diff
@@ -858,6 +861,17 @@ func (s *WindowsService) connectRDP(ctx context.Context, log *slog.Logger, tdpCo
+		computerName = strings.Split(desktop.GetAddr(), ":")[0]
+	}
+
+	kdcAddr := s.cfg.KDCAddr
+	if kdcAddr == "" {
+		kdcAddr = strings.Split(s.cfg.LDAPConfig.Addr, ":")[0]
+	}
```
Same here, use a function built for parsing network addresses rather than a generic string function.

Also consider using `kdcAddr := cmp.Or(kdcAddr, ...)`.

probakowski replied:
I went with `utils.Host` and changed the condition, `cmp.Or` won't work now

zmb3 reviewed (commented):
A couple questions:

- does this also work in FIPS mode?
- did you verify that local users still work with these changes?

Lastly, I think it would be good to get a draft PR for docs open before we merge this. A lot of times you can identify UX papercuts by writing the docs ahead of time.

probakowski committed:
Update lib/srv/desktop/rdp/rdpclient/client_common.go

Co-authored-by: Zac Bergquist <zac.bergquist@goteleport.com>

probakowski committed:
cleanup

probakowski committed:
cleanup

probakowski committed:
cleanup

probakowski committed:
Merge remote-tracking branch 'origin/master' into probakowski/nla

probakowski committed:
cleanup

probakowski commented on Cargo.toml:
There's some delay in `sspi` release so I moved PR to IronRDP repo and just updated hash (as we've done previously during initial IronRDP development)

probakowski replied:
`sspi` was released, we now use head of `master` branch of `IronRDP`

probakowski commented:
@zmb3 I verified local users, everything looks fine there, I'll try to test FIPS mode tomorrow but I don't foresee any problems

probakowski committed:
update sspi to 0.13

probakowski committed:
disable credssp in FIPS builds

probakowski committed:
Merge remote-tracking branch 'origin/master' into probakowski/nla

probakowski committed:
update IronRDP

probakowski committed:
propagate KDCAddr from config file

probakowski committed:
Allow tokio-rustls license

probakowski commented on .github/workflows/dependency-review.yaml:
```diff
@@ -21,6 +21,7 @@ jobs:
         pkg:cargo/ring,
         pkg:cargo/sspi,
         pkg:cargo/tokio-boring,
+        pkg:cargo/tokio-rustls,
```
`tokio-rustls` uses `MIT/Apache-2.0` but dependecy review action can't handle slashes

probakowski replied:
Some yes (`curve25519-dalek-derive`), some should work with vanilla dependency review rules (most are MIT or Apache). I'll revisit it in separate PR - we want to merge this one ASAP

probakowski committed:
clippy

probakowski commented:
@zmb3 @codingllama @tcsc friendly ping

probakowski committed:
Fix public key data in FIPS mode, enable CredSSP in FIPS

zmb3 commented on lib/srv/desktop/rdp/rdpclient/src/rdpdr.rs:
```diff
@@ -167,6 +168,7 @@ impl TeleportRdpdrBackend {
 }
 
 /// A generic error type for the TeleportRdpdrBackend that can contain any arbitrary error message.
+#[allow(dead_code)]
```
Why do we need this. Is this truly unused code?

probakowski replied:
It's false positive, `clippy` complaints about unused `String` but it's used in derived `Debug`

codingllama replied:
Could we explain with a comment?

Same for all others.

zmb3 commented on lib/srv/desktop/rdp/rdpclient/src/rdpdr.rs:
```diff
@@ -79,12 +79,13 @@ impl RdpdrBackend for TeleportRdpdrBackend {
     }
 
-    fn handle_drive_io_request(&mut self, req: ServerDriveIoRequest) -> PduResult<()> {
+    fn handle_drive_io_request(&mut self, req: ServerDriveIoRequest) -> PduResult<Vec<SvcMessage>> {
```
Why do we need to change the return type here?

probakowski replied:
`RdpdrBackend` trait was changed in current version of IronRDP

zmb3 commented on lib/srv/desktop/rdp/rdpclient/src/ssl.rs:
```diff
@@ -56,7 +56,9 @@ pub(crate) async fn upgrade(
         let public_key = cert.public_key()?;
-        let bytes = public_key.public_key_to_der()?;
+        let mut bytes: Vec<u8> = public_key.public_key_to_der()?;
+        //boring uses additional DER element before raw key data compared to rustls, so we have to skip it
```
```suggestion
        // boring uses additional DER element before raw key data compared to rustls, so we have to skip it
```

zmb3 commented on lib/srv/desktop/rdp/rdpclient/src/network_client.rs:
```diff
@@ -0,0 +1,154 @@
+                NetworkProtocol::Tcp => self.send_tcp(&request.url, &request.data).await,
+                NetworkProtocol::Udp => self.send_udp(&request.url, &request.data).await,
+                NetworkProtocol::Http | NetworkProtocol::Https => {
+                    self.send_http(&request.url, &request.data).await
```
Do we ever need this client to send HTTP data? Would it be better to error out or do nothing in this code path?

probakowski replied:
No we don't. I simplify the client, now we error out for HTTP, HTTPS and UDP

zmb3 commented on lib/srv/desktop/rdp/rdpclient/src/network_client.rs:
```diff
@@ -0,0 +1,154 @@
+        Ok(buf)
+    }
+
+    async fn send_udp(&self, url: &Url, data: &[u8]) -> ConnectorResult<Vec<u8>> {
```
`send_udp` is not the best name here, because this does a send and a receive operation.

probakowski replied:
I simplify the client, now we error out for HTTP, HTTPS and UDP

zmb3 commented on lib/srv/desktop/rdp/rdpclient/src/network_client.rs:
```diff
@@ -0,0 +1,154 @@
+    }
+
+    async fn send_udp(&self, url: &Url, data: &[u8]) -> ConnectorResult<Vec<u8>> {
+        let udp_socket = UdpSocket::bind((IpAddr::V4(Ipv4Addr::LOCALHOST), 0))
```
Is `LOCALHOST` okay here?

probakowski replied:
I simplify the client, now we error out for HTTP, HTTPS and UDP

probakowski committed:
Update lib/srv/desktop/rdp/rdpclient/src/ssl.rs

Co-authored-by: Zac Bergquist <zac.bergquist@goteleport.com>

probakowski committed:
simplify network client

codingllama commented on Cargo.toml:
```diff
@@ -20,15 +20,15 @@ codegen-units = 1
-ironrdp-svc = { git = "https://github.com/Devolutions/IronRDP", rev = "dfbe947e5b9bd2da06f1e14620ee4d68bca4252f" }
-ironrdp-tls = { git = "https://github.com/Devolutions/IronRDP", rev = "dfbe947e5b9bd2da06f1e14620ee4d68bca4252f", features = ["rustls" ] }
-ironrdp-tokio = { git = "https://github.com/Devolutions/IronRDP", rev = "dfbe947e5b9bd2da06f1e14620ee4d68bca4252f" }
+ironrdp-cliprdr = { git = "https://github.com/Devolutions/IronRDP", rev = "92efe2adf7402c15fe6cf2da0d3f8ff8ebd767c3" }
```
Could you mention why the updates are necessary?

codingllama commented on lib/srv/desktop/rdp/rdpclient/src/ssl.rs:
```diff
@@ -56,7 +56,9 @@ pub(crate) async fn upgrade(
-        let bytes = public_key.public_key_to_der()?;
+        let mut bytes: Vec<u8> = public_key.public_key_to_der()?;
+        // boring uses additional DER element before raw key data compared to rustls, so we have to skip it
+        bytes.drain(0..24);
```
Is this guaranteed to always be the same length?

codingllama reviewed (commented):
Go looks reasonable to me. As for Rust, you can find much better reviewers than myself.

Could you please clarify the terms in the PR description? You spelled out "AD", but I don't know what "NLA" is.

zmb3 committed:
Minor improvements

- Better error display for IronRDP errors with poor Display impls
- Check key length before truncating to avoid a panic
- Improve godocs

zmb3 committed:
Make NLA opt in via TELEPORT_ENABLE_RDP_NLA=yes

zmb3 committed:
Suppress sspi-rs logs

sspi-rs currently logs everything at info level, which clutters
the Teleport logs. Set its log level to warn until upstream's
logging is less chatty.

zmb3 commented on lib/srv/desktop/rdp/rdpclient/src/client.rs:
```diff
@@ -1469,14 +1512,30 @@ impl Display for ClientError {
+                    write!(f, "CredSSP {:?}: {}", e.error_type, e.description)
+                }
+                ConnectorErrorKind::Custom => {
+                    write!(f, "CredSSP: {}", e.context)?;
```
Reminder to self to fix - might not be a CredSSP error

zmb3 replied:
```suggestion
                    write!(f, "Error: {}", e.context)?;
```

probakowski committed:
try disabling credssp first

probakowski committed:
Merge remote-tracking branch 'origin/master' into probakowski/nla

probakowski committed:
use missing computer name in case of IP address

probakowski committed:
always enable credssp when opted-in

probakowski committed:
fmt

zmb3 commented on lib/srv/desktop/windows_server.go:
```diff
@@ -858,6 +862,24 @@ func (s *WindowsService) connectRDP(ctx context.Context, log *slog.Logger, tdpCo
+	computerName, ok := desktop.GetLabel(types.DiscoveryLabelWindowsDNSHostName)
+	if !ok {
+		if computerName, err = utils.Host(desktop.GetAddr()); err != nil {
+			return trace.Wrap(err)
```
Can we add a more user-friendly message to this error?

probakowski replied:
Added

zmb3 commented on lib/srv/desktop/windows_server.go:
```diff
@@ -858,6 +862,24 @@ func (s *WindowsService) connectRDP(ctx context.Context, log *slog.Logger, tdpCo
+		if computerName, err = utils.Host(desktop.GetAddr()); err != nil {
+			return trace.Wrap(err)
+		}
+		if len(net.ParseIP(computerName)) != 0 {
```
Let's add a comment explaining why we're doing this.

zmb3 commented on lib/srv/desktop/windows_server.go:
```diff
@@ -858,6 +862,24 @@ func (s *WindowsService) connectRDP(ctx context.Context, log *slog.Logger, tdpCo
+	kdcAddr := s.cfg.KDCAddr
+	if kdcAddr == "" && s.cfg.LDAPConfig.Addr != "" {
+		if kdcAddr, err = utils.Host(s.cfg.LDAPConfig.Addr); err != nil {
+			return trace.Wrap(err)
```
Add a message to the error. Something like

```
trace.Wrap(err, "KDC address is unspecified and LDAP address is invalid")
```

Also, should this be a fatal error? If we're connecting to a non-AD host we don't need to error out here, right?

probakowski replied:
I've added check to only do this for AD desktops

probakowski committed:
Update lib/srv/desktop/rdp/rdpclient/src/client.rs

Co-authored-by: Zac Bergquist <zac.bergquist@goteleport.com>

probakowski committed:
review comments

probakowski committed:
Merge remote-tracking branch 'origin/master' into probakowski/nla

probakowski committed:
review comments

probakowski committed:
use teleport.dev/computer_name

probakowski committed:
e

probakowski committed:
Merge branch 'master' into probakowski/nla

probakowski commented on e:
yep, cleaned that up
